src/main/java/teammates/sqllogic/core/CoursesLogic.java

```diff
@@ -50,4 +53,66 @@ public Course createCourse(Course course) throws InvalidParametersException, Ent
     public Course getCourse(String courseId) {
         return coursesDb.getCourse(courseId);
     }
+
+    /**
+     * Deletes a course and cascade its students, instructors, sessions, responses, deadline extensions and comments.
+     * Fails silently if no such course.
+     */
+    public void deleteCourseCascade(String courseId) {
+        if (getCourse(courseId) == null) {
+            return;
+        }
+
+        // TODO: Migrate after other Logic classes have been migrated.
+        // AttributesDeletionQuery query = AttributesDeletionQuery.builder()
+        //         .withCourseId(courseId)
+        //         .build();
+        // frcLogic.deleteFeedbackResponseComments(query);
+        // frLogic.deleteFeedbackResponses(query);
+        // fqLogic.deleteFeedbackQuestions(query);
+        // feedbackSessionsLogic.deleteFeedbackSessions(query);
+        // studentsLogic.deleteStudents(query);
+        // instructorsLogic.deleteInstructors(query);
+        // deadlineExtensionsLogic.deleteDeadlineExtensions(query);
+
+        coursesDb.deleteCourse(courseId);
+    }
+
+    /**
+     * Moves a course to Recycle Bin by its given corresponding ID.
+     * @return the time when the course is moved to the recycle bin.
+     */
+    public Instant moveCourseToRecycleBin(String courseId) throws EntityDoesNotExistException {
+        return coursesDb.softDeleteCourse(courseId);
+    }
+
+    /**
+     * Restores a course from Recycle Bin by its given corresponding ID.
+     */
+    public void restoreCourseFromRecycleBin(String courseId) throws EntityDoesNotExistException {
+        coursesDb.restoreDeletedCourse(courseId);
+    }
+
+    /**
+     * Updates a course by {@link CourseAttributes.UpdateOptions}.
+     *
+     * <p>If the {@code timezone} of the course is changed, cascade the change to its corresponding feedback sessions.
+     *
+     * @return updated course
+     * @throws InvalidParametersException if attributes to update are not valid
+     * @throws EntityDoesNotExistException if the course cannot be found
+     */
+    public Course updateCourseCascade(Course course)
+            throws InvalidParametersException, EntityDoesNotExistException {
+        Course oldCourse = coursesDb.getCourse(course.getId());
+        Course updatedCourse = coursesDb.updateCourse(course);
```

Comment:
Let's stick to this flow for updating courses.

1. Get the course in the api layer. This ensures we fetch a managed entity. You can check also that it exists this way.
2. Use setters to set the updatable fields on the entity. Since we only use POST req, we should be able to directly set the (validated) updated fields. 
3. Call updateCourse on the new entity. This ensures the validity checks and done.

I have some concern about separation of concerns here, but it's no worse than what we currently have. We can try and see how it goes, refactor later if it gets out of hand.